src/server/transaction.cc

```diff
@@ -798,24 +801,38 @@ void Transaction::ExecuteAsync() {
   auto cb = [seq, this] {
     EngineShard* shard = EngineShard::tlocal();
 
+    // this callback should fail from time to time with non-atomic txs in debug mode because we
+    // clear the shard_data_ array.
+    DCHECK(!this->shard_data_.empty());
+
+    // TODO: to document why we need seqlock_. I suspect I added it to guard shard_data
+    // for multi-command transactions. Specifically, if we resize shard_data_ by coordinator thread
+    // and this callback is executed by engine shard thread, we may access invalid memory.
+    // I think the approach is still broken because we may read seqlock_ and pause, then
+    // the coordinator will progress, exit the WaitForShardCallbacks and will resize shard_data_
+    // and then this callback will call IsArmedInShard and will have invalid memory access.
     uint32_t seq_after = seqlock_.load(memory_order_acquire);
-    bool should_poll = (seq_after == seq) && (GetLocalMask(shard->shard_id()) & ARMED);
-
-    DVLOG(3) << "PollExecCb " << DebugId() << " sid(" << shard->shard_id() << ") "
-             << run_count_.load(memory_order_relaxed) << ", should_poll: " << should_poll;
-
-    // We verify that this callback is still relevant.
-    // If we still have the same sequence number and local_mask is ARMED it means
-    // the coordinator thread has not crossed WaitForShardCallbacks barrier.
-    // Otherwise, this callback is redundant. We may still call PollExecution but
-    // we should not pass this to it since it can be in undefined state for this callback.
-    if (should_poll) {
-      // shard->PollExecution(this) does not necessarily execute this transaction.
-      // Therefore, everything that should be handled during the callback execution
-      // should go into RunInShard.
-      shard->PollExecution("exec_cb", this);
-    } else {
-      VLOG(1) << "Skipping PollExecution " << DebugId() << " sid(" << shard->shard_id() << ")";
+    if (seq_after == seq) {
+      bool is_armed = IsArmedInShard(shard->shard_id());
+
+      DVLOG(3) << "PollExecCb " << DebugId() << " sid(" << shard->shard_id() << ") "
+               << run_count_.load(memory_order_relaxed);
+
+      // We verify that this callback is still relevant.
+      // If we still have the same sequence number and tx is armed, meaning
+      // the coordinator thread has not crossed WaitForShardCallbacks barrier (WRONG!).
+      // Otherwise, this callback is redundant. We may still call PollExecution but
+      // we should not pass this to it since it can be in undefined state for this callback.
+      // TODO: I am pretty sure we can achieve both seqlock and refcount semantics
+      // with a single run_count_ atomic.
```

Comment:
Partially, you can introduce "epochs", so that run_cnt value will be `epoch * 1000 + real run_cnt`.
Otherwise, you miss the case where coordinator re-arms for second hop and you'll miss the 0 from the old callback